exercises/binary-search-tree/example.h

```diff
@@ -0,0 +1,226 @@
+#if !defined(_BINARY_SEARCH_TREE_H_)
+#define _BINARY_SEARCH_TREE_H_
+
+#include <memory>
+#include <cstdint>
+
+namespace binary_tree
+{
+    template<typename T>
+    class binary_tree final
+    {
+    public:
+        class binary_tree_iter;
+        using binary_tree_ptr = std::unique_ptr<binary_tree>;
+
+        explicit binary_tree(T data)
+            : _data(std::move(data)),
+              _left(nullptr),
+              _right(nullptr)
+        {}
+
+        ~binary_tree() = default;
+
+        binary_tree(binary_tree &) = delete;
+        binary_tree& operator=(binary_tree &) = delete;
+
+        binary_tree(binary_tree &&) = delete;
+        binary_tree&& operator=(binary_tree &&) = delete;
+
+    public:
+        void insert(const T &data);
+
+        const T &data() const {return _data;};
+        const binary_tree_ptr& left() const {return _left;};
+        const binary_tree_ptr& right() const {return _right;};
+
+        binary_tree_iter begin() const;
+        binary_tree_iter end() const;
+        
+    private:
+        void insert(const T &data, binary_tree_ptr &location);
+        T _data;
+        binary_tree_ptr _left;
+        binary_tree_ptr _right;
+
+    public:
+        class binary_tree_iter final
+        {
+            friend binary_tree_iter binary_tree::begin() const;
+            friend binary_tree_iter binary_tree::end() const;
+        public:
+            using binary_tree_iter_ptr = std::unique_ptr<binary_tree_iter>;
+        private:
+            enum class state
+            {
+                Left,
+                Right,
+                Middle,
+                Done
+            };
+
+            const binary_tree &_tree;
+            state _state;
+            binary_tree_iter_ptr _side_iter;
```

Comment:
The name `_side_iter` was very confusing for me. Maybe `_branch_iter` would be better?